### ISSUE TITLE
`unicorn/no-declarations-before-early-exit` false positive with post-await cancellation guard

### ISSUE THREAD

**Issue by seitenca** [sindresorhus/eslint-plugin-unicorn#3298]
I think `unicorn/no-declarations-before-early-exit` is a bit too eager around post-`await` guards.

The rule currently reports this kind of code:

```js
async function poll(stopSignal) {
	const status = await getStatus();
	if (stopSignal()) {
		return;
	}

	updateUi(status);
}
```

The suggested move is:

```js
async function poll(stopSignal) {
	if (stopSignal()) {
		return;
	}

	const status = await getStatus();
	updateUi(status);
}
```

But those two snippets are not equivalent. In the first one, `stopSignal()` is checked after `getStatus()` resolves. That matters when the stop state can change while the async work is pending.

This came up in a polling callback where `stopSignal()` is backed by reactive state. The intent is:

1. fetch the latest status;
2. check whether polling was stopped while the request was in flight;
3. only update the UI if it is still active.

Moving the declaration after the guard checks the stop state too early, and the UI can still be updated after cancellation.

To preserve the same behavior after moving the declaration, the code would need another guard:

```js
async function poll(stopSignal) {
	if (stopSignal()) {
		return;
	}

	const status = await getStatus();
	if (stopSignal()) {
		return;
	}

	updateUi(status);
}
```

So I think the rule should probably skip declarations whose initializer contains `await`, at least when the early-exit condition is a call expression or otherwise not known to be stable.

Rule: `unicorn/no-declarations-before-early-exit`
Version: `eslint-plugin-unicorn@66.0.0`
